staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go

```diff
@@ -83,29 +83,79 @@ func (g *Cloud) ToInstanceReferences(zone string, instanceNames []string) (refs
 }
 
 // NodeAddresses is an implementation of Instances.NodeAddresses.
-func (g *Cloud) NodeAddresses(_ context.Context, _ types.NodeName) ([]v1.NodeAddress, error) {
-	internalIP, err := metadata.Get("instance/network-interfaces/0/ip")
-	if err != nil {
-		return nil, fmt.Errorf("couldn't get internal IP: %v", err)
-	}
-	externalIP, err := metadata.Get("instance/network-interfaces/0/access-configs/0/external-ip")
-	if err != nil {
-		return nil, fmt.Errorf("couldn't get external IP: %v", err)
-	}
-	addresses := []v1.NodeAddress{
-		{Type: v1.NodeInternalIP, Address: internalIP},
-		{Type: v1.NodeExternalIP, Address: externalIP},
-	}
+func (g *Cloud) NodeAddresses(ctx context.Context, nodeName types.NodeName) ([]v1.NodeAddress, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+	defer cancel()
+
+	if g.isCurrentInstance(string(nodeName)) { // Try to use metadata server first
 
-	if internalDNSFull, err := metadata.Get("instance/hostname"); err != nil {
-		klog.Warningf("couldn't get full internal DNS name: %v", err)
-	} else {
-		addresses = append(addresses,
-			v1.NodeAddress{Type: v1.NodeInternalDNS, Address: internalDNSFull},
-			v1.NodeAddress{Type: v1.NodeHostName, Address: internalDNSFull},
-		)
+		nics, err := metadata.Get("instance/network-interfaces/")
+		if err != nil {
+			return nil, fmt.Errorf("couldn't get network interfaces: %v", err)
+		}
+
+		nodeAddresses := []v1.NodeAddress{}
+		for _, nic := range nics {
+			internalIP, err := metadata.Get("instance/network-interfaces/" + string(nic) + "ip")
+			if err != nil {
+				return nil, fmt.Errorf("couldn't get internal IP: %v", err)
+			}
+			nodeAddresses = append(nodeAddresses,v1.NodeAddress{Type: v1.NodeInternalIP, Address: internalIP})
+
+			acs, err := metadata.Get("instance/network-interfaces/" + string(nic) + "access-configs/")
+			if err != nil {
+				return nil, fmt.Errorf("couldn't get access configs: %v", err)
+			}
+
+			for _, ac := range acs {
+				externalIP, err := metadata.Get("instance/network-interfaces/" + string(nic) + "access-configs/" + string(ac) + "external-ip")
+				if err != nil {
+					return nil, fmt.Errorf("couldn't get external IP: %v", err)
+				}
+				
+				if externalIP != "" {
+					nodeAddresses = append(nodeAddresses,v1.NodeAddress{Type: v1.NodeExternalIP, Address: externalIP})
+				}
+			}
+		}
+
+		internalDNSFull, err := metadata.Get("instance/hostname")
+		if err != nil {
+			klog.Warningf("couldn't get full internal DNS name: %v", err)
+		} else {
+			nodeAddresses = append(nodeAddresses,
+				v1.NodeAddress{Type: v1.NodeInternalDNS, Address: internalDNSFull},
+				v1.NodeAddress{Type: v1.NodeHostName, Address: internalDNSFull},
+			)
+		}
+		return nodeAddresses, nil
+
+	} else { // Use GCE API
```

Comment:
Linter should complain about unnecessary `else` because there is a return statement just above.

Reply:
How? That return is inside the if block

Reply:
Thanks, changed it